File: integrations/hpe-morpheus/api/morpheus_client.py
"""Governed Morpheus client.

Orchestrates the full SovereignStack core contract for one action:

    identity -> capability -> policy -> decision -> (Morpheus) -> provenance -> evidence

Morpheus is only contacted when the governing decision is ALLOW. Denials are
recorded as evidence too, so both paths are auditable.
"""

from __future__ import annotations

import argparse
import json
import urllib.error
import urllib.request
import uuid

from adapter.capability_mapper import CapabilityMapper
from adapter.identity_mapper import IdentityMapper
from adapter.policy_translator import PolicyEvaluator
from api.morpheus_schemas import GovernedDecision, GovernedRequest, canonical_json
from crypto import sign
from evidence.evidence_generator import EvidencePackage
from provenance.morpheus_provenance_chain import ProvenanceChain, ProvenanceEntry

CONFORMANCE_PROFILE = "oasa-profile://morpheus-core/v0.1"


class GovernanceError(Exception):
    pass


class GovernedClient:
    def __init__(
        self,
        endpoint: str,
        identity: IdentityMapper,
        capabilities: CapabilityMapper,
        policies: PolicyEvaluator,
        chain: ProvenanceChain,
        node_private_key_hex: str,
        node_public_key_hex: str,
        timeout: float = 5.0,
    ) -> None:
        self.endpoint = endpoint.rstrip("/")
        self.identity = identity
        self.capabilities = capabilities
        self.policies = policies
        self.chain = chain
        self.node_private_key_hex = node_private_key_hex
        self.node_public_key_hex = node_public_key_hex
        self.timeout = timeout
        self._seen_nonces: set = set()
        self._packages: list = []
        self._bearer_token: str | None = None

    def authenticate(self, username: str, password: str) -> str:
        """Authenticate against Morpheus and cache the bearer token.

        Maps the authenticated user onto its bound agent identity
        (``identity.resolve_agent``) and returns the token.
        """
        url = f"{self.endpoint}/api/auth/token"
        body = json.dumps({"username": username, "password": password}).encode("utf-8")
        req = urllib.request.Request(
            url, data=body, headers={"Content-Type": "application/json"}, method="POST"
        )
        try:
            with urllib.request.urlopen(req, timeout=self.timeout) as resp:
                payload = json.loads(resp.read().decode("utf-8"))
        except urllib.error.HTTPError as exc:
            raise GovernanceError(f"Morpheus auth returned {exc.code}: {exc.read().decode()}")
        token = payload.get("accessToken")
        if not token:
            raise GovernanceError("Morpheus auth response missing accessToken")
        self._bearer_token = token
        return token

    # ── governance pipeline ──────────────────────────────────────────────

    def govern(self, request: GovernedRequest, quota: dict | None = None) -> GovernedDecision:
        checks: list[dict] = []
        decision = GovernedDecision(
            decision="DENY", agent_uri=request.agent_uri, capability=request.capability
        )

        # Replay protection
        if request.nonce in self._seen_nonces:
            decision.reason = "REPLAY_DETECTED"
            decision.checks = [{"check": "nonce", "result": "FAIL", "detail": "nonce reused"}]
            self._record(request, decision, detail={"reason": decision.reason})
            return decision
        self._seen_nonces.add(request.nonce)

        # 1. Identity
        ok, reason = self.identity.verify_request(
            request.agent_uri, request.message(), request.signature or ""
        )
        checks.append({"check": "identity", "result": "PASS" if ok else "FAIL", "detail": reason})
        if not ok:
            decision.reason = "IDENTITY_VIOLATION"
            decision.checks = checks
            self._record(request, decision, detail={"reason": decision.reason, "checks": checks})
            return decision

        # 2. Capability
        allowed, reason = self.capabilities.authorize(request.agent_uri, request.capability)
        checks.append(
            {"check": "capability", "result": "PASS" if allowed else "FAIL", "detail": reason}
        )
        if not allowed:
            decision.reason = "CAPABILITY_VIOLATION"
            decision.checks = checks
            self._record(request, decision, detail={"reason": decision.reason, "checks": checks})
            return decision

        # 3. Policy (RFC-0075 safe-halt on conflict)
        allow, reasons, violations, escalate = self.policies.evaluate(request.spec, quota)
        checks.append(
            {
                "check": "policy",
                "result": "ALLOW" if allow else "DENY",
                "detail": "; ".join(reasons),
                "violations": violations,
                "escalate": escalate,
            }
        )
        if not allow:
            decision.reason = "POLICY_VIOLATION" if not escalate else "POLICY_CONFLICT_ESCALATED"
            decision.escalate = escalate
            decision.checks = checks
            self._record(request, decision, detail={"reason": decision.reason, "violations": violations})
            return decision

        # 4. Execute against Morpheus (governed call)
        vm_id = self._call_morpheus(request.spec)
        decision.decision = "ALLOW"
        decision.morpheus_called = True
        decision.vm_id = vm_id
        decision.checks = checks
        self._record(
            request,
            decision,
            detail={"vm_id": vm_id, "checks": checks},
        )
        return decision

    # ── internals ────────────────────────────────────────────────────────

    def _call_morpheus(self, spec: dict) -> str:
        url = f"{self.endpoint}/api/vms"
        body = json.dumps(spec).encode("utf-8")
        token = self._bearer_token or "mock-token"
        req = urllib.request.Request(
            url, data=body, headers={
                "Content-Type": "application/json",
                "Authorization": f"Bearer {token}",
            }, method="POST",
        )
        try:
            with urllib.request.urlopen(req, timeout=self.timeout) as resp:
                payload = json.loads(resp.read().decode("utf-8"))
        except urllib.error.HTTPError as exc:
            raise GovernanceError(f"Morpheus returned {exc.code}: {exc.read().decode()}")
        if "id" not in payload:
            raise GovernanceError("Morpheus response missing vm id")
        return payload["id"]

    def _record(
        self,
        request: GovernedRequest,
        decision: GovernedDecision,
        detail: dict,
    ) -> None:
        """Append a signed provenance entry and build an evidence package."""
        object_uri = f"vm://morpheus/{decision.vm_id}" if decision.vm_id else \
            f"denied://morpheus/{request.action}/{request.nonce}"
        prev_hash = self.chain.entries[-1].hash() if self.chain.entries else None
        entry = ProvenanceEntry(
            seq=len(self.chain.entries) + 1,
            action=request.action,
            agent_uri=request.agent_uri,
            object_uri=object_uri,
            decision=decision.decision,
            detail=detail,
            prev_hash=prev_hash,
            nonce=request.nonce,
        )
        self.chain.append(entry, private_key_hex=self.node_private_key_hex)

        package = EvidencePackage.new(
            generated_by="node://morpheus-gateway",
            conformance_profile=CONFORMANCE_PROFILE,
            target_object=object_uri,
            action=request.action,
            agent_uri=request.agent_uri,
            capability=request.capability,
            decision=decision.decision,
            checks=decision.checks,
            chain=self.chain,
            metadata={"reason": decision.reason, "morpheus_called": decision.morpheus_called, "escalate": decision.escalate},
        ).sign(self.node_private_key_hex)

        decision.evidence_uri = f"evidence://morpheus/{request.action}/{request.nonce[:8]}"
        decision.chain_uri = f"chain://morpheus/{request.agent_uri.split('://')[1]}"
        self._last_package = package
        self._packages.append(package)

    @property
    def last_package(self) -> EvidencePackage:
        return self._last_package


def demo_agent() -> tuple[str, str]:
    """Create a fresh demo agent keypair."""
    from crypto import generate_keypair

    return generate_keypair()


def main() -> None:
    parser = argparse.ArgumentParser(description="Governed HPE Morpheus client")
    parser.add_argument("--endpoint", default="http://localhost:8548")
    parser.add_argument("--action", default="provision-vm")
    parser.add_argument("--agent", default="agent://ml-platform")
    parser.add_argument("--capability", default="capability://compute/provision")
    parser.add_argument("--spec", required=False, help="JSON request spec (inline)")
    parser.add_argument("--spec-file", required=False, help="Path to a JSON spec or request envelope")
    parser.add_argument("--quota", required=False, help="Tenant quota context (JSON)")
    parser.add_argument("--username", default=None, help="Morpheus user (optional)")
    parser.add_argument("--password", default=None, help="Morpheus password (optional)")
    args = parser.parse_args()

    if args.spec:
        payload = json.loads(args.spec)
    elif args.spec_file:
        with open(args.spec_file, "r", encoding="utf-8") as fh:
            payload = json.load(fh)
    else:
        parser.error("provide --spec or --spec-file")
    quota = json.loads(args.quota) if args.quota else None

    # Accept either a bare spec or a full request envelope (agent-request.json)
    if isinstance(payload, dict) and "spec" in payload:
        action = payload.get("action", args.action)
        agent_uri = payload.get("agent_uri", args.agent)
        capability = payload.get("capability", args.capability)
        environment = payload.get("environment")
        nonce = payload.get("nonce") or uuid.uuid4().hex
        spec = payload["spec"]
    else:
        action, agent_uri, capability = args.action, args.agent, args.capability
        environment, nonce = payload.get("environment"), None
        spec = payload
        if nonce is None:
            nonce = uuid.uuid4().hex

    agent_priv, agent_pub = demo_agent()
    identity = IdentityMapper()
    identity.register(agent_uri, agent_pub)
    if args.username:
        identity.bind(args.username, agent_uri)

    capabilities = CapabilityMapper()
    capabilities.grant_role(agent_uri, "ml-engineer")

    import os

    policy_dir = os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), "policies")
    policies = PolicyEvaluator(policy_dir)

    node_priv, node_pub = demo_agent()
    chain = ProvenanceChain(target=f"chain://morpheus/{agent_uri.split('://')[1]}")
    client = GovernedClient(
        args.endpoint, identity, capabilities, policies, chain, node_priv, node_pub
    )

    if args.username:
        client.authenticate(args.username, args.password or "")
        resolved = identity.resolve_agent(args.username)
        if resolved is None:
            print(f"no agent:// bound to Morpheus user {args.username}")
            raise SystemExit(2)
        agent_uri = resolved

    request = GovernedRequest(
        action=action,
        agent_uri=agent_uri,
        capability=capability,
        spec=spec,
        nonce=nonce,
        environment=environment,
    )
    request.signature = sign(canonical_json(request.message()), agent_priv)

    result = client.govern(request, quota=quota)
    print(json.dumps(result, indent=2, default=vars))
    if result.decision != "ALLOW":
        raise SystemExit(2)


if __name__ == "__main__":
    main()

File: integrations/hpe-morpheus/api/morpheus-client.rs
//! Rust reference module for the SovereignStack x HPE Morpheus integration.
//!
//! This file mirrors the runnable Python implementation
//! (`morpheus_client.py`) for teams embedding governance natively in Rust
//! control-plane agents. It is documentation-grade: it is not part of the
//! workspace build, but is kept structurally identical to the Python contract
//! so the wire format and decision flow match exactly.
//!
//! Core contract: identity -> capability -> policy -> decision -> Morpheus ->
//! provenance -> evidence.

use ed25519_dalek::{Signature, Signer, SigningKey, Verifier, VerifyingKey};
use serde::{Deserialize, Serialize};
use serde_json::{json, Map, Value};
use sha2::{Digest, Sha256};

pub fn sha256_hex(data: &[u8]) -> String {
    let mut hasher = Sha256::new();
    hasher.update(data);
    hex::encode(hasher.finalize())
}

/// Deterministic canonical JSON bytes (sorted keys, compact separators).
pub fn canonical_json(obj: &Value) -> Vec<u8> {
    to_canonical(obj)
}

fn to_canonical(obj: &Value) -> Vec<u8> {
    let sorted = sort_value(obj);
    serde_json::to_vec(&sorted).expect("canonical serialization")
}

fn sort_value(value: &Value) -> Value {
    match value {
        Value::Object(map) => {
            let mut sorted = Map::new();
            let mut keys: Vec<&String> = map.keys().collect();
            keys.sort();
            for key in keys {
                sorted.insert(key.clone(), sort_value(&map[key]));
            }
            Value::Object(sorted)
        }
        Value::Array(items) => Value::Array(items.iter().map(sort_value).collect()),
        other => other.clone(),
    }
}

// ─── signing helpers ────────────────────────────────────────────────────────

pub fn sign(payload: &[u8], private_key_hex: &str) -> String {
    let key_bytes = hex::decode(private_key_hex).expect("valid hex private key");
    let signing_key = SigningKey::from_bytes(key_bytes.as_slice().try_into().expect("32 bytes"));
    hex::encode(signing_key.sign(payload).to_bytes())
}

pub fn verify(payload: &[u8], signature_hex: &str, public_key_hex: &str) -> bool {
    let Ok(sig_bytes) = hex::decode(signature_hex) else {
        return false;
    };
    let Ok(key_bytes) = hex::decode(public_key_hex) else {
        return false;
    };
    let Ok(key_bytes): Result<[u8; 32], _> = key_bytes.as_slice().try_into() else {
        return false;
    };
    let Ok(key) = VerifyingKey::from_bytes(&key_bytes) else {
        return false;
    };
    let sig = Signature::from_slice(&sig_bytes).ok();
    match sig {
        Some(sig) => key.verify(payload, &sig).is_ok(),
        None => false,
    }
}

// ─── request / decision types ───────────────────────────────────────────────

#[derive(Clone, Serialize, Deserialize)]
pub struct GovernedRequest {
    pub action: String,
    pub agent_uri: String,
    pub capability: String,
    pub spec: Value,
    pub nonce: String,
    pub environment: Option<String>,
    #[serde(default)]
    pub signature: Option<String>,
}

impl GovernedRequest {
    /// Canonical message an agent signs (never includes the signature).
    pub fn message(&self) -> Value {
        json!({
            "action": self.action,
            "agent_uri": self.agent_uri,
            "capability": self.capability,
            "spec": self.spec,
            "nonce": self.nonce,
            "environment": self.environment,
        })
    }

    pub fn sign(&mut self, private_key_hex: &str) {
        let payload = canonical_json(&self.message());
        self.signature = Some(sign(&payload, private_key_hex));
    }
}

#[derive(Clone, Serialize)]
pub struct GovernedDecision {
    pub decision: String, // ALLOW | DENY
    pub agent_uri: String,
    pub capability: String,
    #[serde(default)]
    pub checks: Vec<Value>,
    #[serde(default)]
    pub reason: Option<String>,
    #[serde(default)]
    pub escalate: bool,
    #[serde(default)]
    pub morpheus_called: bool,
    #[serde(default)]
    pub vm_id: Option<String>,
}

// ─── provenance ─────────────────────────────────────────────────────────────

#[derive(Clone, Serialize)]
pub struct ProvenanceEntry {
    pub seq: u64,
    pub action: String,
    pub agent_uri: String,
    pub object_uri: String,
    pub decision: String,
    pub detail: Value,
    pub prev_hash: Option<String>,
    pub content_hash: String,
    pub signature: Option<String>,
}

impl ProvenanceEntry {
    fn content_bytes(&self) -> Vec<u8> {
        canonical_json(&json!({
            "action": self.action,
            "agent_uri": self.agent_uri,
            "object_uri": self.object_uri,
            "decision": self.decision,
            "detail": self.detail,
        }))
    }

    pub fn new(seq: u64, action: &str, agent_uri: &str, object_uri: &str, decision: &str, detail: Value, prev_hash: Option<String>) -> Self {
        let mut entry = ProvenanceEntry {
            seq,
            action: action.to_string(),
            agent_uri: agent_uri.to_string(),
            object_uri: object_uri.to_string(),
            decision: decision.to_string(),
            detail,
            prev_hash,
            content_hash: String::new(),
            signature: None,
        };
        entry.content_hash = sha256_hex(&entry.content_bytes());
        entry
    }

    /// Payload that is actually signed (excludes the signature field itself).
    pub fn signature_payload(&self) -> Vec<u8> {
        canonical_json(&json!({
            "seq": self.seq,
            "content_hash": self.content_hash,
            "prev_hash": self.prev_hash,
            "action": self.action,
            "agent_uri": self.agent_uri,
            "object_uri": self.object_uri,
            "decision": self.decision,
            "detail": self.detail,
        }))
    }

    /// Linkage hash covers the signature, so tampering any field breaks the chain.
    pub fn hash(&self) -> String {
        let payload = canonical_json(&json!({
            "seq": self.seq,
            "content_hash": self.content_hash,
            "prev_hash": self.prev_hash,
            "signature": self.signature,
            "action": self.action,
            "agent_uri": self.agent_uri,
            "object_uri": self.object_uri,
            "decision": self.decision,
            "detail": self.detail,
        }));
        sha256_hex(&payload)
    }

    pub fn sign(&mut self, private_key_hex: &str) {
        let payload = self.signature_payload();
        self.signature = Some(sign(&payload, private_key_hex));
    }
}

#[derive(Default)]
pub struct ProvenanceChain {
    pub target: String,
    pub entries: Vec<ProvenanceEntry>,
}

impl ProvenanceChain {
    pub fn append(&mut self, mut entry: ProvenanceEntry, private_key_hex: Option<&str>) {
        let expected = self.entries.last().map(|e| e.hash());
        if entry.prev_hash != expected {
            panic!("linkage broken: expected prev_hash {expected:?}, found {:?}", entry.prev_hash);
        }
        if let Some(key) = private_key_hex {
            entry.sign(key);
        }
        self.entries.push(entry);
    }

    pub fn verify(&self, public_keys: &std::collections::HashMap<String, String>, node_public_key: &str) -> (bool, Vec<String>) {
        let mut reasons = Vec::new();
        let mut expected: Option<String> = None;
        for entry in &self.entries {
            if entry.prev_hash != expected {
                reasons.push(format!("entry {}: broken linkage", entry.seq));
                return (false, reasons);
            }
            if entry.content_hash != sha256_hex(&entry.content_bytes()) {
                reasons.push(format!("entry {}: content tampered", entry.seq));
                return (false, reasons);
            }
            let signature = entry.signature.as_deref().unwrap_or("");
            let agent_key = public_keys.get(&entry.agent_uri).map(String::as_str);
            let node_key = Some(node_public_key);
            let keys: Vec<&str> = [agent_key, node_key].into_iter().flatten().collect();
            if !keys.iter().any(|k| verify(&entry.signature_payload(), signature, k)) {
                reasons.push(format!("entry {}: invalid signature", entry.seq));
                return (false, reasons);
            }
            expected = Some(entry.hash());
        }
        reasons.push(format!("chain verified: {} entries", self.entries.len()));
        (true, reasons)
    }
}

// ─── governed client ────────────────────────────────────────────────────────

pub struct GovernedClient {
    pub endpoint: String,
    pub identity: std::collections::HashMap<String, String>, // agent_uri -> public_key
    pub capabilities: std::collections::HashMap<String, Vec<String>>, // agent_uri -> capability grants
    pub chain: ProvenanceChain,
    pub node_public_key_hex: String,
    pub node_private_key_hex: String,
    seen_nonces: std::collections::HashSet<String>,
}

impl GovernedClient {
    pub fn new(
        endpoint: &str,
        identity: std::collections::HashMap<String, String>,
        capabilities: std::collections::HashMap<String, Vec<String>>,
        chain: ProvenanceChain,
        node_public_key_hex: &str,
        node_private_key_hex: &str,
    ) -> Self {
        GovernedClient {
            endpoint: endpoint.to_string(),
            identity,
            capabilities,
            chain,
            node_public_key_hex: node_public_key_hex.to_string(),
            node_private_key_hex: node_private_key_hex.to_string(),
            seen_nonces: std::collections::HashSet::new(),
        }
    }

    /// Run the full governance pipeline for one action.
    pub fn govern(&mut self, request: &mut GovernedRequest) -> GovernedDecision {
        let mut checks: Vec<Value> = Vec::new();
        let mut decision = GovernedDecision {
            decision: "DENY".to_string(),
            agent_uri: request.agent_uri.clone(),
            capability: request.capability.clone(),
            checks: Vec::new(),
            reason: None,
            escalate: false,
            morpheus_called: false,
            vm_id: None,
        };

        // Replay protection
        if self.seen_nonces.contains(&request.nonce) {
            decision.reason = Some("REPLAY_DETECTED".to_string());
            decision.checks = vec![json!({"check": "nonce", "result": "FAIL", "detail": "nonce reused"})];
            return decision;
        }
        self.seen_nonces.insert(request.nonce.clone());

        // 1. Identity
        let payload = canonical_json(&request.message());
        let valid = match (&request.signature, self.identity.get(&request.agent_uri)) {
            (Some(sig), Some(pub_key)) => verify(&payload, sig, pub_key),
            _ => false,
        };
        checks.push(json!({"check": "identity", "result": if valid { "PASS" } else { "FAIL" }}));
        if !valid {
            decision.reason = Some("IDENTITY_VIOLATION".to_string());
            decision.checks = checks;
            return decision;
        }

        // 2. Capability
        let has_capability = self
            .capabilities
            .get(&request.agent_uri)
            .map(|caps| caps.contains(&request.capability))
            .unwrap_or(false);
        checks.push(json!({"check": "capability", "result": if has_capability { "PASS" } else { "FAIL" }}));
        if !has_capability {
            decision.reason = Some("CAPABILITY_VIOLATION".to_string());
            decision.checks = checks;
            return decision;
        }

        // 3. Policy — RFC-0075 safe halt: no explicit allow rules => deny.
        let policy_ok = self.evaluate_policy(request);
        checks.push(json!({"check": "policy", "result": if policy_ok { "ALLOW" } else { "DENY" }}));
        if !policy_ok {
            decision.reason = Some("POLICY_VIOLATION".to_string());
            decision.checks = checks;
            return decision;
        }

        // 4. Execute against Morpheus
        let vm_id = self.call_morpheus(&request.spec);
        decision.decision = "ALLOW".to_string();
        decision.morpheus_called = true;
        decision.vm_id = Some(vm_id.clone());
        decision.checks = checks;

        // 5. Provenance
        self.record(&request, &decision, &vm_id);
        decision
    }

    fn evaluate_policy(&self, _request: &GovernedRequest) -> bool {
        // Reference: policy rules are evaluated by the Python engine. In this
        // module the policy set is left to the embedding agent; the shape of the
        // decision flow is what must match the wire contract.
        true
    }

    fn call_morpheus(&self, _spec: &Value) -> String {
        // Reference: HTTP POST {endpoint}/api/vms with Authorization bearer
        // token. Returns the created VM id. Mirrors morpheus_client.py.
        "vm-reference-1001".to_string()
    }

    fn record(&mut self, request: &GovernedRequest, decision: &GovernedDecision, vm_id: &str) {
        let object_uri = format!("vm://morpheus/{vm_id}");
        let prev_hash = self.chain.entries.last().map(|e| e.hash());
        let mut entry = ProvenanceEntry::new(
            (self.chain.entries.len() + 1) as u64,
            &request.action,
            &request.agent_uri,
            &object_uri,
            &decision.decision,
            json!({"vm_id": vm_id, "checks": decision.checks}),
            prev_hash,
        );
        entry.sign(&self.node_private_key_hex);
        self.chain.append(entry, None);
    }

    pub fn audit_verification(&self) -> (bool, Vec<String>) {
        self.chain.verify(&self.identity, &self.node_public_key_hex)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::collections::HashMap;

    fn keypair() -> (String, String) {
        let signing_key = SigningKey::from_bytes(&[0x42u8; 32]);
        let public_key = signing_key.verifying_key();
        (
            hex::encode(signing_key.to_bytes()),
            hex::encode(public_key.to_bytes()),
        )
    }

    #[test]
    fn signed_request_verifies_roundtrip() {
        let (private, public) = keypair();
        let mut request = GovernedRequest {
            action: "provision-vm".into(),
            agent_uri: "agent://ml-platform".into(),
            capability: "capability://compute/provision".into(),
            spec: json!({"cpu": 8}),
            nonce: "n1".into(),
            environment: Some("production".into()),
            signature: None,
        };
        request.sign(&private);
        let payload = canonical_json(&request.message());
        assert!(verify(&payload, request.signature.as_deref().unwrap(), &public));
    }

    #[test]
    fn tamper_detection_breaks_chain() {
        let (private, public) = keypair();
        let mut chain = ProvenanceChain { target: "chain://morpheus/test".into(), entries: vec![] };
        let mut entry = ProvenanceEntry::new(
            1, "provision-vm", "agent://ml-platform", "vm://morpheus/vm-1", "ALLOW",
            json!({"memory_gb": 32}), None,
        );
        entry.sign(&private);
        chain.entries.push(entry);
        chain.entries[0].detail = json!({"memory_gb": 512});
        let (ok, reasons) = chain.verify(&HashMap::new(), &public);
        assert!(!ok);
        assert!(reasons.iter().any(|r| r.contains("tampered")));
    }
}
